lib/http/request/writer.rb

```
@@ -51,12 +51,16 @@ def connect_through_proxy
       # Adds the headers to the header array for the given request body we are working
       # with
       def add_body_type_headers
-        if @body.is_a?(String) && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
-        elsif @body.nil? && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: 0"
-        elsif @body.is_a?(Enumerable) && CHUNKED != @headers[Headers::TRANSFER_ENCODING]
-          raise(RequestError, "invalid transfer encoding")
+        unless @headers[Headers::CONTENT_LENGTH]
+          if @body.is_a?(String)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
+          elsif @body.respond_to?(:read)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.size}"
```

Oh. I got what you mean. I guess then we should raise an exception:

``` ruby
elsif @body.respond_to? :read
  raise "IO object must respond to #size" unless @body.respond_to? :size
  # ...
```